### PR TITLE
Hqt

### PR DESCRIPTION
**Pull request by bgoldberg-habana**
enable Habana quantization flow for fp8 inference.
add an option to use kv cache in lm_eval
wrap kv cache in module for HQT

### CODE REVIEW

File: examples/text-generation/README.md [hunk 1]
```diff
@@ -98,8 +98,8 @@ Here are a few settings you may be interested in:
 - `--prompt` to benchmark the model on one or several prompts of your choice
 - `--attn_softmax_bf16` to run attention softmax layer in bfloat16 precision provided that the model (such as Llama) supports it
 - `--trim_logits` to calculate logits only for the last token in the first time step provided that the model (such as Llama) supports it
-- `--kv_cache_fp8` Store kv-cache in float8 when kv-cache is used
 - `--fp8` Enable Quantization to fp8
+- `--kv_cache_fp8` Deprecated - Store kv-cache in float8 when kv-cache is used. should not be used with HQT(Habana Quantization Toolkit)
```
Comment (MrGeva) [inline examples/text-generation/README.md:102]: per business, we should not use "Habana Quantization Toolkit". we agreed to use "The Quantization Toolkit". this is aligned with RTD

**Review by MrGeva — CHANGES_REQUESTED** [review]

File: examples/text-generation/README.md [hunk 2]
```diff
@@ -218,36 +218,67 @@ python run_generation.py \
 Another way to simulate dynamic input is to use `--simulate_dyn_prompt`. For example `--simulate_dyn_prompt 25,35,45` will extend or crop the default prompt (or the prompt passed in using `--prompt`) to sizes 25, 35, and 45, and throughput will be measured for these 3 lengths. If `--simulate_dyn_prompt` is used, the min and max input lengths from it are computed to perform warmup as well. One final optimization that can be used in case of dynamic inputs is `--reduce_recompile`. Thus the suggested configuration to simulate dynamicity after warmup is to use all three arguments: `--simulate_dyn_prompt 25 35 45 --reduce_recompile --bucket_size 30`
 ### Running with FP8
 
-Llama2-7b in FP8 is enabled. Use `--fp8` to enable quantization in fp8.
-Add the `--kv_cache_fp8` argument to run the model with a KV cache allocated in fp8.
+Llama2-70b and Llama2-7b in FP8 are enabled using the Quantization Toolkit (HQT), which provides model measurement and quantization capabilities in PyTorch.
+
 More information on enabling fp8 in SynapseAI is available here:
 https://docs.habana.ai/en/latest/PyTorch/Inference_on_PyTorch/Inference_Using_FP8.html
```
Comment (MrGeva) [inline examples/text-generation/README.md:224]: need to make sure the page is updated for 1.14 first. currently its not merged yet by the TR

**Review by MrGeva — CHANGES_REQUESTED** [review]

File: examples/text-generation/README.md [hunk 3]
```diff
@@ -218,36 +218,67 @@ python run_generation.py \
...
+--trim_logits \
+--use_kv_cache \
+--reuse_cache \
+--bf16 \
+--batch_size 277 \
+--max_new_tokens 2048 \
+--max_input_tokens 2048 \
+--limit_hpu_graphs \
+--fp8
+```
```
[the remark this answers is not in the document]
Comment (libinta) [reply examples/text-generation/README.md:273]: it uses different scripts

File: examples/text-generation/README.md (hunk 2 above)
Comment (libinta) [reply examples/text-generation/README.md:224]: we are in the processing of update, it sould be done in a day or 2.

File: examples/text-generation/quantization_config/act_maxabs_hw_weights_pcs_maxabs_pow2_quant.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+    "method": "HOOKS",
+    "mode": "QUANTIZE",
+    "observer": "maxabs",
+    "scale_method": "ACT_MAXABS_HW_WEIGHTS_PCS_MAXABS_POW2",
+    "whitelist": {"types": [], "names":  []},
+    "blacklist": {"types": [], "names":  []},
+    "dump_stats_path": "./hqt_output/measure",
+    "dump_stats_xlsx_path": "./hqt_output/measure/fp8stats.xlsx"
```
[the remark this answers is not in the document]
Comment (libinta) [reply examples/text-generation/quantization_config/act_maxabs_hw_weights_pcs_maxabs_pow2_quant.json:9]: the cmd given is to use maxabs_quant.json (below) to generate the xslx file

File: examples/text-generation/README.md (hunk 2 above)
Comment (libinta) [reply examples/text-generation/README.md:224]: Done

File: examples/text-generation/utils.py [hunk 5]
```diff
@@ -178,7 +183,7 @@ def setup_distributed_model(args, model_dtype, model_kwargs, logger):
 
     logger.info("DeepSpeed is enabled.")
     deepspeed.init_distributed(dist_backend="hccl")
-    config = AutoConfig.from_pretrained(args.model_name_or_path, **model_kwargs)
+    config = AutoConfig.from_pretrained(args.model_name_or_path, torch_dtype=model_dtype, **model_kwargs)
```
[the remark this answers is not in the document]
Comment (libinta) [reply examples/text-generation/utils.py:186]: modeling_llama.py, dtype = self.config.torch_dtype, if we don't change it, we got fp16, not bf16

**Review by libinta — APPROVED** [review]

**Review by libinta — APPROVED** [review]

**Pull request merged**